crates/shared/src/price_estimation/competition.rs

```diff
@@ -53,32 +56,37 @@ type PriceEstimationStage<T> = Vec<(String, T)>;
 pub struct RacingCompetitionEstimator<T> {
     inner: Vec<PriceEstimationStage<T>>,
     successful_results_for_early_return: NonZeroUsize,
+    ranking: PriceRanking,
 }
 
 impl<T: Send + Sync + 'static> RacingCompetitionEstimator<T> {
     pub fn new(
         inner: Vec<PriceEstimationStage<T>>,
         successful_results_for_early_return: NonZeroUsize,
+        ranking: PriceRanking,
     ) -> Self {
         assert!(!inner.is_empty());
         Self {
             inner,
             successful_results_for_early_return,
+            ranking,
         }
     }
 
     fn estimate_generic<
         Q: Clone + Debug + Send + 'static,
         R: Clone + Debug + Send,
         E: Clone + Debug + Send,
+        C,
     >(
         &self,
         query: Q,
         kind: OrderKind,
         get_single_result: impl Fn(&T, Q) -> futures::future::BoxFuture<'_, Result<R, E>>
             + Send
             + 'static,
-        compare_results: impl Fn(&Result<R, E>, &Result<R, E>) -> Ordering + Send + 'static,
+        compare_results: impl Fn(&Result<R, E>, &Result<R, E>, &C) -> Ordering + Send + 'static,
+        provide_additional_context: impl FusedFuture<Output = Result<C, E>> + Send + 'static,
```

Comment:
Does this have to be so generic? Don't we always return a concrete `RankingContext` ?

Reply:
We don't have any context in the generic price estimation [case](https://github.com/cowprotocol/services/pull/2145/files#diff-bf5c6f66e24401168ed4c2c36453fad49c5ee0f7ff88e6c845490bdeea5605bcR223) because it doesn't appear to make sense to give the user the option between `MaxOutAmount` and `BestBangForBuck` because for `BestBangForBuck` we need to have a native price which is what we are trying to compute in that call in the first case.
So unless I'm missing something it's not really possible to decide in that case whether it makes sense to use a less complex trade with lower `out_amount` because we don't really know yet how the token relates to native `ETH`.